elephant/spike_train_surrogates.py

```diff
@@ -144,13 +140,95 @@ def dither_spikes(spiketrain, dither, n=1, decimals=None, edges=True):
+    """
+    Generates surrogates of a spike train by uniform dithering conserving the
+    refractory period.
+
+    The surrogates are obtained by uniformly dithering times around the
+    original position. The dithering is performed independently for each
+    surrogate. The dithering range is for each spike adjusted such, that it
+    conserves the refractory period.
+
+    The surrogates retain the :attr:`t_start` and :attr:`t_stop` of the
+    original `SpikeTrain` object.
+
+    Parameters
+    ----------
+    spiketrain : neo.SpikeTrain
+        The spiketrain from which to generate the surrogates from.
+    dither : pq.Quantity
+        Amount of dithering.
+        To a spike a uniform random step within the window
+        '[-previous_dither, next_dither]' is added. With:
+        'previous_dither = min(dither,
+            spike - previous_spike - refractory_period)'
+        'next_dither = min(dither,
+            next_spike - spike - refractory_period)'
+    n_surrogates : int, optional
+        Number of surrogates to create.
+        Default: 1
+    refractory_period
+        The starting 'refractory_period' to work with.
+        If 'min(isi) < refractory_period', function works with 'min(isi)'
+        as 'refractory_period'.
+        Default: 4.*pq.ms
+
+    Returns
+    -------
+    dithered_spiketrains : list of neo.SpikeTrain
+        Each element is an independent surrogate spiketrain of 'spiketrain',
+        where each spike is moved.
+    """
+    if len(spiketrain) == 0:
+        return [spiketrain] * n_surrogates
+
+    unit = spiketrain.units
+    t_start, t_stop = spiketrain.t_start.magnitude, spiketrain.t_stop.magnitude
+    spiketrain = spiketrain.magnitude
+
+    dither = dither.rescale(unit).magnitude
+    refractory_period = refractory_period.rescale(unit).magnitude
+    refractory_period = np.min(np.diff(spiketrain), initial=refractory_period)
+
+    dithered_spiketrains = []
+    for _ in range(n_surrogates):
+        dithered_spiketrain = np.copy(spiketrain)
+        random_ordered_ids = np.random.choice(
+            len(spiketrain), size=len(spiketrain), replace=False)
+
+        for random_id in random_ordered_ids:
+            spike = dithered_spiketrain[random_id]
+
+            previous_spike = dithered_spiketrain[random_id - 1] \
+                if random_id > 0 \
+                else t_start - refractory_period
+            next_spike = dithered_spiketrain[random_id + 1] \
+                if random_id < len(spiketrain) - 1 \
+                else t_stop + refractory_period
+
+            previous_dither = min(dither,
+                                  spike - previous_spike - refractory_period)
+            next_dither = min(dither,
+                              next_spike - spike - refractory_period)
+
+            dithered_spiketrain[random_id] += (previous_dither + next_dither)\
+                * random.random() - previous_dither
```

Comment:
I'm sure we can vectorize the dithering and get rid of the internal for-loop. Try adding a random uniform vector, things like that.
Think of how to do this. We can think together.

Reply:
I found a way to vectorize this if the following logic is equivalent to
```python
dithered_spiketrain = np.copy(spiketrain.magnitude)
        for i, random_id in enumerate(random_ordered_ids):
            spike = spiketrain.magnitude[random_id]

            prev_spike = spiketrain.magnitude[random_id - 1] \
                if random_id > 0 \
                else t_start - refractory_period
            next_spike = spiketrain.magnitude[random_id + 1] \
                if random_id < len(spiketrain) - 1 \
                else t_stop + refractory_period

            prev_dither = min(dither,
                                  spike - prev_spike - refractory_period)
            next_dither = min(dither,
                              next_spike - spike - refractory_period)

            dt = (prev_dither + next_dither) * random_vals[i] - prev_dither
            dithered_spiketrain[random_id] += dt
```

but it breaks the test.
So do we need to use `dithered_spiketrain` instead of `spiketrain.magnitude` in the for-loop?

Reply:
I think it breaks because prev_spike and next_spike needs to be taken off the updated spike train, i.e. of dithered_spiketrain.

Reply:
This is the logic 
`dithered_spiketrain = np.copy(spiketrain.magnitude)
        for i, random_id in enumerate(random_ordered_ids):
            spike = spiketrain.magnitude[random_id]
                 % or spike = dithered_spiketrain[random_id]
                 % both forms are equivalent here since every spike is only moved once

            prev_spike = dithered_spiketrain[random_id - 1] \
                if random_id > 0 \
                else t_start - refractory_period
            next_spike = dithered_spiketrain[random_id + 1] \
                if random_id < len(spiketrain) - 1 \
                else t_stop + refractory_period

            prev_dither = min(dither,
                                  spike - prev_spike - refractory_period)
            next_dither = min(dither,
                              next_spike - spike - refractory_period)

            dt = (prev_dither + next_dither) * random.random() - prev_dither
            dithered_spiketrain[random_id] += dt`

Reply:
OK, sad. Then there is no way to vectorize the for loop.

Reply:
Yes, that is was I thought.